### ISSUE TITLE
Explicit use of an index in a for loop instead of .iter().enumerate()

### ISSUE THREAD

**Issue by jdm** [Manishearth/rust-clippy#159]
```rust
let mut index = 0;
for element in collection.iter() {
    ...
    index += 1;
}
```
vs.
```rust
for (index, element) in collection.iter().enumerate() {
}
```
Note that the first loop may also be using `&collection`, which would be worth testing separately presumably.

**Comment by Manishearth** [comment]
(This needs some AST walking to find the `index += 1`; but is otherwise similar to #157)

**Comment by llogiq** [comment]
related to #11

**Comment by nweston** [comment]
I'll take a look at this and #11.

**Comment by Manishearth** [comment]
Sure! Let us know if you need any help!

The easiest thing to do here would be to match on the desugared for loop (which you can get with https://github.com/Manishearth/rust-clippy/blob/65deb956f1f2560ba9fa9148aef5b111bccc02cc/src/loops.rs#L91) and check if its inner block contains an increment statement.

However, you also need to see if the parent block has a `let mut index=0`. You can accomplish this either by writing a visitor that walks function definitions, or by using `get_parent_node` and checking if the parent block (if it is a block) contains an initialization of the same variable to zero followed by no accesses.

**Comment by nweston** [comment]
I have this basically working, but there are some false positives mostly caused by shadowing. For example, this will warn:
```
let mut index = 0;
let mut index = 1;
for v in vec.iter() { index += 1 }
```
I can fix this by continuing to scan past the first definition to make sure it isn't shadowed later on. But this struck me as the kind of problem that might already be solved elsewhere in the compiler. Is there a way, given a variable, to find its definition?

What about finding uses of a variable? That would simplify the implementation of this a lot.

**Comment by Manishearth** [comment]
I think if you do a `def_map` lookup it will give you the node where it was declared.

As for uses, you need to use `ExprUseVisitor`

**Comment by nweston** [comment]
OK, shadowing is taken care of now, but I still have a couple of false positives.

This will warn about both these loops, even though the index variable is no longer initialized to zero at the start of the second loop:
```
let mut index = 0;
for v in vec.iter() { index += 1 }
for v in vec.iter() { index += 1 }
```
And this will warn because my lint doesn't realize that `index` isn't incremented on every iteration:
```
let mut index = 0;
for v in vec.iter() { if *v == 1 { index += 1 } }
```

I'm taking the basic approach suggested by @Manishearth (look for an incremented variable in the loop, then check the parent block to see if it's initialized to zero). Any thoughts on how to weed out these last false positives?

**Comment by llogiq** [comment]
For the first, the correct solution is to `enumerate` a `chain` of the `vec` with itself.

For the second, you need to match on ExprIf statements, and bail out if the index increment is inside one.

**Comment by nweston** [comment]
> For the first, the correct solution is to enumerate a chain of the vec with itself.

But only if the bodies of the two loops are identical. It seems like it would be a lot of work for the lint to suggest that.

**Comment by llogiq** [comment]
True – the `eq_op` lint had code to do that, before it was recently pruned to avert false positives. Not too much work actually (and simple, considering that it was my first lint).

Even so, I'm fine with just bailing out, since this cases should be rare.
